### PR TITLE
'podman cp' copy between host and container

### LINKED ISSUE #613: cp support
FEATURE REQUEST

mount/umount is more powerful than cp, but just as buildah is more powerful than a Dockerfile, you still support a Dockerfile.  cp is such a common task it would be nice if it was supported.

### PR DESCRIPTION
**Pull request by QiWang19**
fix #613  Copy files/folders between a container and the local filesystem
Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

File: cmd/podman/cp.go [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+func cpCmd(c *cli.Context) error {
+	args := c.Args()
+	if len(args) != 2 {
+		return errors.Errorf("you must provide source path and destination path")
+	}
+	srcPath := args[0]
+	destPath := args[1]
+
+	if isContainerPathFormat(srcPath) && isContainerPathFormat(destPath) {
+		return errors.Errorf("please provide only one container working dorectory")
```
Comment (TomSweeneyRedHat) [inline cmd/podman/cp.go]: s/directory/dorectory/

File: cmd/podman/cp.go [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+	runtime, err := libpodruntime.GetRuntime(c)
+	if err != nil {
+		return errors.Wrapf(err, "could not get runtime")
+	}
+	defer runtime.Shutdown(false)
+
+	if os.Geteuid() != 0 {
+		if driver := runtime.GetConfig().StorageConfig.GraphDriverName; driver != "vfs" {
+			// Do not allow to mount a graphdriver that is not vfs if we are creating the userns as part
```
Comment (TomSweeneyRedHat) [inline cmd/podman/cp.go]: nit s/allow to mount/allow mounting of/

File: cmd/podman/cp.go [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+		if err != nil {
+			return err
+		}
+		if became {
+			os.Exit(ret)
+		}
+	}
+
+	if isContainerPathFormat(destPath) {
+		name := strings.Split(destPath, ":")[0]
```
Comment (rhatdan) [inline cmd/podman/cp.go]: Should only do the strings.Split onece and set them to src and dest.  Probably better to do a Splitn2() and then verify that their are two and objects.

File: cmd/podman/cp.go [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+		}
+		return nil
+	}
+	return errors.Errorf("invalid parameters: %s, %s", srcPath, destPath)
+}
+
+func copyBetweenHostAndContainer(name, srcPath, destination string, runtime *libpod.Runtime, isFromHostToContainer bool) error {
+	ctr, err := runtime.LookupContainer(name)
+	if err != nil {
+		return errors.Wrapf(err, "error looking up container %q", name)
```
Comment (rhatdan) [inline cmd/podman/cp.go]: err is already wrapped, so just return err.

File: cmd/podman/cp.go [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+)
+
+var (
+	cpDescription = "Copy files/folders between a container and the local filesystem"
+	cpCommand     = cli.Command{
+		Name:         "cp",
+		Usage:        "Copy files/folders between a container and the local filesystem",
+		Description:  cpDescription,
+		Action:       cpCmd,
+		ArgsUsage:    "CONTAINER:SRC_PATH DEST_PATH | SRC_PATH|- CONTAINER:DEST_PAT",
```
Comment (mheon) [inline cmd/podman/cp.go]: [CONTAINER:]SRC_PATH [CONTAINER:]DEST_PATH

(This loses the nuance that only one can be a container, but I think it's still a lot clearer)

File: cmd/podman/cp.go [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+func copyBetweenHostAndContainer(name, srcPath, destination string, runtime *libpod.Runtime, isFromHostToContainer bool) error {
+	ctr, err := runtime.LookupContainer(name)
+	if err != nil {
+		return errors.Wrapf(err, "error looking up container %q", name)
+	}
+	mountPoint, err := ctr.Mount()
+	if err != nil {
+		return errors.Wrapf(err, "error mounting container %q", ctr.ID())
```
Comment (rhatdan) [inline cmd/podman/cp.go]: Already wrapped, just return err.

File: cmd/podman/cp.go [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+	}
+	user, err := getUser(mountPoint, ctr.User())
+	if err != nil {
+		return err
+	}
+	idmappingopts, err := ctr.IDMappings()
```
Comment (rhatdan) [inline cmd/podman/cp.go]: I don't think ctr.IDMappings ever returns an err, so we should probably change the interface.

File: cmd/podman/cp.go [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "could not get runtime")
+	}
+	defer runtime.Shutdown(false)
+
+	if os.Geteuid() != 0 {
+		if driver := runtime.GetConfig().StorageConfig.GraphDriverName; driver != "vfs" {
+			// Do not allow to mount a graphdriver that is not vfs if we are creating the userns as part
+			// of the mount command.
+			return fmt.Errorf("cannot mount using driver %s in rootless mode", driver)
```
Comment (mheon) [inline cmd/podman/cp.go]: Change mount to cp

File: cmd/podman/cp.go [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+
+	if isContainerPathFormat(srcPath) && isContainerPathFormat(destPath) {
+		return errors.Errorf("please provide only one container working dorectory")
+	}
+	if !isContainerPathFormat(srcPath) && !isContainerPathFormat(destPath) {
+		return errors.Errorf("you must provade one container ID")
+	}
+
+	if os.Geteuid() != 0 {
+		rootless.SetSkipStorageSetup(true)
```
Comment (mheon) [inline cmd/podman/cp.go]: On the rootless code here - can we possibly make this work with rootless by leaking the file descriptor for whatever we want to copy into the userns? Wouldn't work for folders, but might be enough to let us copy stuff into rootless containers without using vfs. @giuseppe

File: cmd/podman/cp.go [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+		name := strings.Split(destPath, ":")[0]
+		destination := strings.Split(destPath, ":")[1]
+		err := copyBetweenHostAndContainer(name, srcPath, destination, runtime, true)
+		if err != nil {
+			return errors.Wrapf(err, "error copy from %s to %s", srcPath, destPath)
+		}
+		return nil
+	}
+
+	if isContainerPathFormat(srcPath) {
```
Comment (mheon) [inline cmd/podman/cp.go]: This should be an else-if - we only ever execute one of these

File: cmd/podman/cp.go [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+		name := strings.Split(srcPath, ":")[0]
```
Comment (mheon) [inline cmd/podman/cp.go]: This should be a `SplitN(srcPath, ":", 1)` - `:` is a completely valid character in Linux filenames

File: cmd/podman/cp.go [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+		err := copyBetweenHostAndContainer(name, srcPath, destination, runtime, false)
+		if err != nil {
+			return errors.Wrapf(err, "error copy from %s to %s", srcPath, destPath)
+		}
+		return nil
+	}
+	return errors.Errorf("invalid parameters: %s, %s", srcPath, destPath)
+}
+
+func copyBetweenHostAndContainer(name, srcPath, destination string, runtime *libpod.Runtime, isFromHostToContainer bool) error {
```
Comment (mheon) [inline cmd/podman/cp.go]: Should check that srcPath and destination are len > 0 immediately so we don't try a copy with garbage data

File: cmd/podman/cp.go [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+		err := copyBetweenHostAndContainer(name, srcPath, destination, runtime, true)
+		if err != nil {
+			return errors.Wrapf(err, "error copy from %s to %s", srcPath, destPath)
+		}
+		return nil
+	}
+
+	if isContainerPathFormat(srcPath) {
+		name := strings.Split(srcPath, ":")[0]
+		srcPath = strings.Split(srcPath, ":")[1]
```
Comment (mheon) [inline cmd/podman/cp.go]: We need to check that strings.Split() returned an array of len(2)

File: cmd/podman/cp.go (hunk 7 above)
Comment (mheon) [reply cmd/podman/cp.go]: I think it might return nil sometimes, though - do we need to check if it's non-nil?

Comment (mheon) [reply cmd/podman/cp.go]: Camelcase the idmappingsopts identifier - `idMappingOpts`

File: cmd/podman/cp.go [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+
+	hostOwner := idtools.IDPair{UID: int(hostUID), GID: int(hostGID)}
+
+	var dest string
+	if isFromHostToContainer {
+		dest = mountPoint
+		if destination != "" && filepath.IsAbs(destination) {
+			dest = filepath.Join(dest, destination)
+		} else {
+			if err = idtools.MkdirAllAndChownNew(filepath.Join(dest, ctr.WorkingDir()), 0755, hostOwner); err != nil {
```
Comment (mheon) [inline cmd/podman/cp.go]: Why are we using WorkingDir here? Shouldn't this be MointPoint + destination path?

File: cmd/podman/cp.go [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+	if err != nil {
+		return err
+	}
+
+	hostOwner := idtools.IDPair{UID: int(hostUID), GID: int(hostGID)}
+
+	var dest string
+	if isFromHostToContainer {
+		dest = mountPoint
+		if destination != "" && filepath.IsAbs(destination) {
```
Comment (mheon) [inline cmd/podman/cp.go]: I don't really understand the logic here - do we never create the directory is the path is absolute? Why?

Comment (mheon) [reply cmd/podman/cp.go]: Also, I think we need to make sure that any symlinks here resolve in the scope of the container - @baude what was the library we used for that, again?

File: cmd/podman/cp.go [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+			}
+			if err = idtools.MkdirAllAndChownNew(filepath.Join(dir, dest), 0755, hostOwner); err != nil {
+				return errors.Wrapf(err, "error creating directory %q", filepath.Join(dir, dest))
+			}
+			dest = filepath.Join(dir, dest)
+		}
+
+	}
+
+	if len(destination) > 0 && destination[len(destination)-1] == os.PathSeparator {
```
Comment (mheon) [inline cmd/podman/cp.go]: What's the os.PathSeparator check here? Isn't this the same as `filepath.IsAbs()`?

Comment (mheon) [reply cmd/podman/cp.go]: Oh, is this a check to see if we're copying into the given directory, versus into a directory with a specific name?

File: cmd/podman/cp.go [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+	if !isFromHostToContainer {
+		src := mountPoint
+		if srcPath != "" && filepath.IsAbs(srcPath) {
+			srcPath = filepath.Join(src, srcPath)
+		} else {
+			srcPath = filepath.Join(src, ctr.WorkingDir(), srcPath)
+		}
+	}
+
+	glob, err := filepath.Glob(srcPath)
```
Comment (mheon) [inline cmd/podman/cp.go]: I'm pretty sure this is wrong - per https://docs.docker.com/engine/reference/commandline/cp/ `docker cp` does not support globs. I think we should duplicate the functionality initially, we can think about expanding on it more.

**Comment by mheon** [comment]
I think we need to make sure the cases listed in https://docs.docker.com/engine/reference/commandline/cp/ right after the line beginning with `Assuming a path separator` are properly handled

File: cmd/podman/cp.go [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "invalid glob %q", srcPath)
+	}
+	if len(glob) == 0 {
+		return errors.Wrapf(syscall.ENOENT, "no files found matching %q", srcPath)
+	}
+
+	for _, gsrc := range glob {
+		esrc, err := filepath.EvalSymlinks(gsrc)
```
Comment (mheon) [inline cmd/podman/cp.go]: Need to make sure this is scoped into the container if source is container

**Comment by mheon** [comment]
Also, I'd love more comments, the code is very hard to follow in places right now

File: cmd/podman/cp.go (hunk 16 above)
Comment (QiWang19) [reply cmd/podman/cp.go]: yes...from what I learned from buildah copy :thought_balloon:

File: cmd/podman/cp.go (hunk 9 above)
Comment (giuseppe) [reply cmd/podman/cp.go]: I think we'd need to join the user namespace, once we have joined the user+mount namespace we can also access the storage.

@QiWang19 would you like to take care of it in this PR or would you prefer me to work on it as a followup patch?

Basically we need to use `rootless.JoinDirectUserAndMountNS()`

Comment (QiWang19) [reply cmd/podman/cp.go]: @giuseppe sure 👌 , I prefer you work on that

Comment (giuseppe) [reply cmd/podman/cp.go]: could we just add an error for now? `cp` should fail when running in rootless mode

File: cmd/podman/cp.go (hunk 11 above)
Comment (QiWang19) [reply cmd/podman/cp.go]: `SplitN(srcPath, ":", 2)` ?

File: cmd/podman/cp.go (hunk 3 above)
Comment (QiWang19) [reply cmd/podman/cp.go]:
```go
destPathArr, destIsValidPath := isContainerPathFormat(destPath)
if destIsValidPath {
	name := destPathArr[0]
	destination := destPathArr[1]
        ...
}
```  
 isContainerPathFormat does SplitN and returns the slice and bool 
```go
func isContainerPathFormat(path string) ([]string, bool) {
	pathArr := strings.SplitN(path, ":", 2)
	if len(pathArr) == 2 {
		return pathArr, true
	}
	return nil, false
}
```

File: cmd/podman/cp.go (hunk 10 above)
Comment (QiWang19) [reply cmd/podman/cp.go]: Can I return at the end of last `if {}` code block?

File: cmd/podman/cp.go (hunk 11 above)
Comment (mheon) [reply cmd/podman/cp.go]: Oops - yeah, 2

File: cmd/podman/cp.go (hunk 10 above)
Comment (mheon) [reply cmd/podman/cp.go]: Yeah, if the if and else-if both end in returns, you don't need a separate return at the end

File: cmd/podman/cp.go (hunk 13 above)
Comment (QiWang19) [reply cmd/podman/cp.go]: Can I check here ```func isContainerPathFormat(path string) ([]string, bool) ``` ?

**Comment by QiWang19** [comment]
`libpod/docs/podman-cp.1.md`, how to change the current description :neutral_face:

File: cmd/podman/cp.go [hunk 19]
```diff
@@ -0,0 +1,341 @@
...
+	hostUID, hostGID, err := getHostIDs(idMappingOpts.UIDMap, idMappingOpts.GIDMap, int(user.UID), int(user.GID))
+	if err != nil {
+		return err
+	}
+
+	hostOwner := idtools.IDPair{UID: int(hostUID), GID: int(hostGID)}
+
+	var glob []string
+	if isFromHostToCtr {
+		glob, _ = filepath.Glob(srcPath)
```
Comment (rhatdan) [inline cmd/podman/cp.go]: Only need to do this once outside of the if else block

File: cmd/podman/cp.go [hunk 20]
```diff
@@ -0,0 +1,341 @@
...
+			}
+			destPath = filepath.Join(mountPoint, ctr.WorkingDir(), destPath)
+		}
+	} else {
+		if filepath.IsAbs(srcPath) {
+			srcPath = filepath.Join(mountPoint, srcPath)
+		} else {
+			srcPath = filepath.Join(mountPoint, ctr.WorkingDir(), srcPath)
+		}
+		glob, _ = filepath.Glob(srcPath)
```
Comment (rhatdan) [inline cmd/podman/cp.go]: Move this block below the if/else block.

File: cmd/podman/cp.go [hunk 21]
```diff
@@ -0,0 +1,341 @@
...
+			}
+		} else {
+			u.AdditionalGids = groups
+		}
+
+	}
+	return u, err
+}
+
+func getHostIDs(uidmap, gidmap []idtools.IDMap, uid, gid int) (int, int, error) {
```
Comment (rhatdan) [inline cmd/podman/cp.go]: Can we user the vendor from buildah, rather then copying?

File: cmd/podman/cp.go [hunk 22]
```diff
@@ -0,0 +1,341 @@
...
+			break
+		}
+	}
+	if !gidMapped {
+		return 0, 0, errors.Errorf("container uses ID mappings, but doesn't map GID %d", gid)
+	}
+	return uid, gid, nil
+}
+
+func copyFileWithTar(idMappingOpts storage.IDMappingOptions, chownOpts *idtools.IDPair, hasher io.Writer) func(src, dest string) error {
```
Comment (rhatdan) [inline cmd/podman/cp.go]: Lets make these vendored from buildah

File: cmd/podman/cp.go [hunk 23]
```diff
@@ -0,0 +1,341 @@
...
+			if err == nil {
+				err = errors.Wrapf(hashError, "error calculating digest of data for %q while copying", dest)
+			}
+			return err
+		}
+	}
+	return archiver.CopyFileWithTar
+}
+
+func copyWithTar(idMappingOpts storage.IDMappingOptions, chownOpts *idtools.IDPair, hasher io.Writer) func(src, dest string) error {
```
Comment (rhatdan) [inline cmd/podman/cp.go]: Vendor from buildah.

File: cmd/podman/cp.go [hunk 24]
```diff
@@ -0,0 +1,341 @@
...
+	if hasher != nil {
+		originalUntar := archiver.Untar
+		archiver.Untar = func(tarArchive io.Reader, dest string, options *archive.TarOptions) error {
+			return originalUntar(io.TeeReader(tarArchive, hasher), dest, options)
+		}
+	}
+	return archiver.CopyWithTar
+}
+
+func untarPath(idMappingOpts storage.IDMappingOptions, chownOpts *idtools.IDPair, hasher io.Writer) func(src, dest string) error {
```
Comment (rhatdan) [inline cmd/podman/cp.go]: Vendor from buildah

File: cmd/podman/cp.go (hunk 22 above)
Comment (QiWang19) [reply cmd/podman/cp.go]:
```go
func (b *Builder) copyFileWithTar(chownOpts *idtools.IDPair, hasher io.Writer) func(src, dest string) error
```
can't use with `b *Builder` :thinking:

**Comment by rhatdan** [comment]
@QiWang19 Containers/storage now has the copyarchive stuff, Please rework this to use those vendors.

**Comment by QiWang19** [comment]
> @QiWang19 Containers/storage now has the copyarchive stuff, Please rework this to use those vendors.

I already used that vendor, currently ok?

**Comment by TomSweeneyRedHat** [comment]
@QiWang19 just make sure the vendor.conf in your sandbox has the same id for containers/storage as upstream does.  If not, you'll want to do a fetch/rebase.

**Comment by mheon** [comment]
I see the vendored changes, but no change to vendor.conf

**Comment by QiWang19** [comment]
@TomSweeneyRedHat  :ok_hand: need to change `master` to `v1.9`

**Comment by rhatdan** [comment]
LGTM
@umohnani8 @vrothberg @mheon @baude @TomSweeneyRedHat PTAL
I think this is ready to merge.

File: docs/podman-cp.1.md [hunk 25]
```diff
@@ -3,20 +3,70 @@
...
+
+You can also use : when specifying paths to a **SRC_PATH** or **DEST_PATH** on a local machine, for example, `file:name.txt`.
+
+If you use a : in a local machine path, you must be explicit with a relative or absolute path, for example:
+	`/path/to/file:name.txt` or `./file:name.txt`
+
+
+## ALTERNATIVES
+
+We have a much stronger capability besides `podman cp` to achieve copy files between host and container.
```
Comment (TomSweeneyRedHat) [inline docs/podman-cp.1.md]: I'm not a big fan of "you" in a man page, but really don't like "We".  To me it's kinda like a nurse saying "It's time for our shot!".  Could we change this to:

"There is a much stronger..."

File: docs/podman-cp.1.md [hunk 26]
```diff
@@ -3,20 +3,70 @@
...
+
+Using standard podman-mount and podman-umount, we can take advantage of the entire linux tool chain, rather
```
Comment (TomSweeneyRedHat) [inline docs/podman-cp.1.md]: "podman -umount, we can take" to "podman-umount takes"

File: docs/podman-cp.1.md (hunk 25 above)
Comment (TomSweeneyRedHat) [reply docs/podman-cp.1.md]: That works for me.

**Comment by QiWang19** [comment]
@mheon PTAL

File: cmd/podman/cp.go [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+var (
+	cpDescription = "Copy files/folders between a container and the local filesystem"
+	cpCommand     = cli.Command{
+		Name:         "cp",
+		Usage:        "Copy files/folders between a container and the local filesystem",
```
Comment (4383) [inline cmd/podman/cp.go]: Why to not reuse `cpDescription` here too? like the line below instead of repeat the same string.

Eventually I guess you need to rewrite your usage with something like:
```go
Usage:      "podman cp [CONTAINER:]SRC_PATH [CONTAINER:]DEST_PATH"
```
Thoughts?

Comment (4383) [reply cmd/podman/cp.go]: After testing your generated output seem good:
```
bin/podman cp -h
NAME:
   podman cp - Copy files/folders between a container and the local filesystem

USAGE:
   podman cp [CONTAINER:]SRC_PATH [CONTAINER:]DEST_PATH

DESCRIPTION:
   Copy files/folders between a container and the local filesystem
```

**Comment by 4383** [comment]
I currently trying this nice feature and I observe that my files was not copied from or to my testing container:
```shell
$ bin/podman ps
CONTAINER ID  IMAGE                            COMMAND  CREATED         STATUS             PORTS  NAMES
e46d67e9477f  docker.io/library/ubuntu:latest  bash     38 seconds ago  Up 38 seconds ago         cocky_swanson 
$ bin/podman exec cocky_swanson touch boum.txt
$ bin/podman exec cocky_swanson ls -la /
total 72
drwxr-xr-x.  21 root   root    4096 Feb  7 11:01 .
drwxr-xr-x.   2 root   root    4096 Feb  4 21:05 bin
drwxr-xr-x.   2 root   root    4096 Apr 24  2018 boot
-rw-r--r--.   1 root   root       5 Feb  7 11:07 boum.txt
drwxr-xr-x.   5 root   root     360 Feb  7 10:50 dev
drwxr-xr-x.   4 root   root    4096 Feb  6 03:37 etc
drwxr-xr-x.   2 root   root    4096 Apr 24  2018 home
drwxr-xr-x.   4 root   root    4096 May 23  2017 lib
drwxr-xr-x.   2 root   root    4096 Feb  4 21:03 lib64
drwxr-xr-x.   2 root   root    4096 Feb  4 21:02 media
drwxr-xr-x.   2 root   root    4096 Feb  4 21:02 mnt
drwxr-xr-x.   2 root   root    4096 Feb  4 21:02 opt
dr-xr-xr-x. 300 nobody nogroup    0 Feb  7 10:50 proc
drwx------.   2 root   root    4096 Feb  4 21:04 root
drwxr-xr-x.   2 root   root    4096 Feb  7 10:50 run
drwxr-xr-x.   2 root   root    4096 Feb  6 03:37 sbin
drwxr-xr-x.   2 root   root    4096 Feb  4 21:02 srv
dr-xr-xr-x.  13 nobody nogroup    0 Feb  5 16:57 sys
drwxrwxrwt.   2 root   root    4096 Feb  4 21:05 tmp
drwxr-xr-x.   7 root   root    4096 Feb  4 21:02 usr
drwxr-xr-x.   2 root   root    4096 Feb  4 21:04 var
$ bin/podman cp $(pwd)/README.md cocky_swanson:/README.md 
$ bin/podman exec cocky_swanson ls -la /   # nothing news was created inside the container
total 72
drwxr-xr-x.  21 root   root    4096 Feb  7 11:01 .
drwxr-xr-x.   2 root   root    4096 Feb  4 21:05 bin
drwxr-xr-x.   2 root   root    4096 Apr 24  2018 boot
-rw-r--r--.   1 root   root       5 Feb  7 11:07 boum.txt
drwxr-xr-x.   5 root   root     360 Feb  7 10:50 dev
drwxr-xr-x.   4 root   root    4096 Feb  6 03:37 etc
drwxr-xr-x.   2 root   root    4096 Apr 24  2018 home
drwxr-xr-x.   4 root   root    4096 May 23  2017 lib
drwxr-xr-x.   2 root   root    4096 Feb  4 21:03 lib64
drwxr-xr-x.   2 root   root    4096 Feb  4 21:02 media
drwxr-xr-x.   2 root   root    4096 Feb  4 21:02 mnt
drwxr-xr-x.   2 root   root    4096 Feb  4 21:02 opt
dr-xr-xr-x. 299 nobody nogroup    0 Feb  7 10:50 proc
drwx------.   2 root   root    4096 Feb  4 21:04 root
drwxr-xr-x.   2 root   root    4096 Feb  7 10:50 run
drwxr-xr-x.   2 root   root    4096 Feb  6 03:37 sbin
drwxr-xr-x.   2 root   root    4096 Feb  4 21:02 srv
dr-xr-xr-x.  13 nobody nogroup    0 Feb  5 16:57 sys
drwxrwxrwt.   2 root   root    4096 Feb  4 21:05 tmp
drwxr-xr-x.   7 root   root    4096 Feb  4 21:02 usr
drwxr-xr-x.   2 root   root    4096 Feb  4 21:04 var
$  bin/podman cp cocky_swanson:/boum.txt /tmp   # error appear during copy
error evaluating symlinks "": lstat /home/herve/.local/share/containers/storage/overlay/df486e0ab8ab9782aae30ce09f181e715aab2981f0e5d1e1735625fcf6b88f07/merged/boum.txt: no such file or directory
```

**Comment by 4383** [comment]
Just FYI (related to my previous message), if I inspect the storage overlay after the copy from the host to the container I can observe that my files is present inside the overlay but not inside the container:
```shell
ls /home/herve/.local/share/containers/storage/overlay/df486e0ab8ab9782aae30ce09f181e715aab2981f0e5d1e1735625fcf6b88f07/merged 
README.md
```

**Comment by QiWang19** [comment]
thanks, working on this. but confusing, looks like I don't meet this issue  :confounded: 

```
drwxr-xr-x   11 root     root          4096 Dec 20 22:25 var
[qiwan@qiwan libpod]$ bin/podman cp $(pwd)/docs/libpod.conf.5.md unruffled_montalcini:/libpod-1.conf.5.md
[qiwan@qiwan libpod]$ bin/podman exec unruffled_montalcini ls -la /
total 92
drwxr-xr-x   18 root     root          4096 Feb  7 15:07 .
drwxr-xr-x   18 root     root          4096 Feb  7 15:07 ..
drwxr-xr-x    2 root     root          4096 Dec 20 22:25 bin
drwxr-xr-x    5 root     root           340 Feb  7 14:50 dev
drwxr-xr-x   15 root     root          4096 Feb  7 14:36 etc
-rw-rw-r--    1 root     root          3858 Jan 11 23:15 foo.md
drwxr-xr-x    2 root     root          4096 Dec 20 22:25 home
-rw-rw-r--    1 root     root          3858 Jan 11 23:15 koo.md
drwxr-xr-x    5 root     root          4096 Dec 20 22:25 lib
-rw-rw-r--    1 root     root          3858 Jan 11 23:15 libpod-1.conf.5.md
-rw-rw-r--    1 root     root          3858 Jan 11 23:15 libpod.conf.5.md
drwxr-xr-x    5 root     root          4096 Dec 20 22:25 media
drwxr-xr-x    2 root     root          4096 Dec 20 22:25 mnt
dr-xr-xr-x  304 nobody   nobody           0 Feb  7 14:50 proc
-rw-rw-r--    1 root     root          3858 Jan 11 23:15 pwdRead.md
-rw-rw-r--    1 root     root          3858 Jan 11 23:15 read.md
-rw-rw-r--    1 root     root          3858 Jan 11 23:15 readd.md
-rw-rw-r--    1 root     root          3858 Jan 11 23:15 red.md
drwx------    2 root     root          4096 Feb  7 14:50 root
drwxr-xr-x    2 root     root          4096 Feb  7 14:36 run
drwxr-xr-x    2 root     root          4096 Dec 20 22:25 sbin
drwxr-xr-x    2 root     root          4096 Dec 20 22:25 srv
dr-xr-xr-x   13 nobody   nobody           0 Feb  7 02:36 sys
drwxrwxrwt    2 root     root          4096 Dec 20 22:25 tmp
drwxr-xr-x    7 root     root          4096 Dec 20 22:25 usr
drwxr-xr-x   11 root     root          4096 Dec 20 22:25 var
```

**Comment by mheon** [comment]
@4383 Is that rootless?

I suspect that rootless `cp` won't work right now - @giuseppe can you confirm?

If so, we should explicitly error on running `cp` as rootless

**Comment by 4383** [comment]
@mheon sorry I'm not sure... I'm a newcomer in this world how I can check this?
@QiWang19 sorry still to doesn't work on my side.

**Comment by mheon** [comment]
@4383 Rootless = running containers as a non-root user

`cp` right now doesn't properly handle the rootless user namespace, I think

**Comment by mheon** [comment]
It helps when I press the right button and don't accidentally close PRs

**Comment by 4383** [comment]
@mheon thanks for the explain :)

So yeah I running containers as a non-root user

**Comment by rhatdan** [comment]
We should through an error right now when they run rootless until @giuseppe Fixes the issues.

File: cmd/podman/cp.go (hunk 27 above)
Comment (QiWang19) [reply cmd/podman/cp.go]: yes, cli.Command.Usage goes to NAME finally :wink:

File: cmd/podman/cp.go [hunk 28]
```diff
@@ -0,0 +1,255 @@
...
+			destPath = filepath.Join(mountPoint, ctr.WorkingDir(), destPath)
+		}
+	} else {
+		if filepath.IsAbs(srcPath) {
+			srcPath = filepath.Join(mountPoint, srcPath)
+		} else {
+			srcPath = filepath.Join(mountPoint, ctr.WorkingDir(), srcPath)
+		}
+	}
+	glob, _ = filepath.Glob(srcPath)
```
Comment (mheon) [inline cmd/podman/cp.go]: Don't drop the error here

File: cmd/podman/cp.go [hunk 29]
```diff
@@ -0,0 +1,255 @@
...
+		if err = os.MkdirAll(destPath, 0755); err != nil {
+			return errors.Wrapf(err, "error creating directory %q", destPath)
+		}
+	} else {
+		if err = os.MkdirAll(filepath.Dir(destPath), 0755); err != nil {
+			return errors.Wrapf(err, "error creating directory %q", destPath)
+		}
+	}
+
+	copyFileWithTar := chrootarchive.CopyFileWithTarAndChown(chownOpts, digest.Canonical.Digester().Hash(), idMappingOpts.UIDMap, idMappingOpts.GIDMap)
```
Comment (mheon) [inline cmd/podman/cp.go]: Maybe a comment here explaining that these return functions for copying items?

File: cmd/podman/cp.go [hunk 30]
```diff
@@ -0,0 +1,255 @@
...
+	srcPath, err := filepath.EvalSymlinks(src)
+	if err != nil {
+		return errors.Wrapf(err, "error evaluating symlinks %q", srcPath)
+	}
+
+	srcPath, srcfi, err := getPathInfo(srcPath)
+	if err != nil {
+		return err
+	}
+	if srcfi.IsDir() || dest[len(dest)-1] == os.PathSeparator {
```
Comment (mheon) [inline cmd/podman/cp.go]: Use strings.HasSuffix() instead of the len(dest)-1 here

**Comment by rhatdan** [comment]
@QiWang19 Lets finish this up tomorrow, I am back in office.

File: cmd/podman/cp.go (hunk 28 above)
Comment (QiWang19) [reply cmd/podman/cp.go]: fixed

File: cmd/podman/cp.go (hunk 29 above)
Comment (QiWang19) [reply cmd/podman/cp.go]: fixed

File: cmd/podman/cp.go (hunk 30 above)
Comment (QiWang19) [reply cmd/podman/cp.go]: fixed

**Comment by rhatdan** [comment]
LGTM

**Comment by QiWang19** [comment]
needs approve label to test tide?

**Comment by mheon** [comment]
Oh, nobody approved
/approve
